### PR TITLE
Fix trailing whitespace and superfluous blank lines in generated GlobalSuppressions.cs

### LINKED ISSUE #37529: Generated GlobalSuppressions.cs has trailing whitespace and two unnecessary lines
**Version Used**: VS 16.2.0

image

Even after cleaning up, any generated addition to the file must be cleaned up too. This is a frequent annoyance:

image

### PR DESCRIPTION
**Pull request by jnm2**
Fixes #37529

Each thing is fixed with a pair of commits, one to add the failing test and the next to make all tests pass again.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs [hunk 1]
```diff
@@ -806,6 +805,96 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
...
+    int Method()
+    {
+        int x = 0;
+    }
+}", expected);
+                }
+
+                private static bool HasTrailingWhitespace(string line)
+                {
+                    return line.LastOrNullable() is { } last && char.IsWhiteSpace(last);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs]: can you just say `is char last`?  seems much more readable.

**Comment by jnm2** [comment]
Ready for review!

**Comment by jnm2** [comment]
Fixed the two distant test failures found by CI in Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGeneration.AddAttributesTests:

- TestAddAssemblyAttributeListToDocumentWithOtherAssemblyAttribute
- TestAddAssemblyAttributeListToEmptyDocument

File: src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs [hunk 2]
```diff
@@ -806,6 +805,94 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
                         new UserDiagnosticAnalyzer(), new CSharpSuppressionCodeFixProvider());
                 }
 
+                [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSuppression)]
+                public async Task GeneratedCodeShouldNotHaveTrailingWhitespace()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs]: 💡 Use `[WorkItem]` to link to the issue

File: src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs [hunk 3]
```diff
@@ -806,6 +805,94 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
...
+}", expected);
+                }
+
+                private static bool HasTrailingWhitespace(string line)
+                {
+                    return line.LastOrNullable() is char last && char.IsWhiteSpace(last);
+                }
+
+                [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSuppression)]
+                public async Task GeneratedCodeShouldNotHaveLeadingBlankLines()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs]: 💡 Use `[WorkItem]` to link to the issue

File: src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs [hunk 4]
```diff
@@ -806,6 +805,94 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
...
+{
+    int Method()
+    {
+        int x = 0;
+    }
+}", expected);
+                }
+
+                [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSuppression)]
+                public async Task GeneratedCodeShouldNotHaveMoreThanOneTrailingBlankLine()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs]: 💡 Use `[WorkItem]` to link to the issue

File: src/Workspaces/CSharp/Portable/Formatting/Rules/ElasticTriviaFormattingRule.cs [hunk 5]
```diff
@@ -276,15 +276,15 @@ private int LineBreaksAfter(SyntaxToken previousToken, SyntaxToken currentToken)
                         || currentToken.Kind() == SyntaxKind.OpenBraceToken) ? 1 : 0;
 
                 case SyntaxKind.CloseBracketToken:
-                    // Assembly and module-level attributes followed by non-attributes should have
-                    // a blank line after them.
+                    // Assembly and module-level attributes followed by non-attributes should have a blank line after
+                    // them, unless it’s the end of the file which will already have a blank line.
```
[the remark this answers is not in the document]
Comment (jnm2) [reply src/Workspaces/CSharp/Portable/Formatting/Rules/ElasticTriviaFormattingRule.cs]: 😢 okay.

**Comment by jnm2** [comment]
@sharwell Thanks, updated.

**Comment by jnm2** [comment]
How do you rerun checks? There is no console output for the 'Build and test' steps except e.g. "The job running on agent NetCorePublic-Pool 67 has exceeded the maximum execution time of 90."

**Comment by mavasani** [comment]
@jnm2 There seems to be legitimate test failure `TestSuppressionWithUsingDirectiveInExistingGlobalSuppressionsDocument`

**Comment by jnm2** [comment]
@mavasani Oh, thank you. I think there was a conflict with my other PR, so this test wouldn't have passed if I had updated it before the other one merged.

**Comment by jnm2** [comment]
@mavasani Oh, thank you. This test didn't exist before my other PR merged to master, so I'll rebase.

**Pull request merged**